controller/disaster_revocery_test.go

```diff
@@ -0,0 +1,166 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/rancher/longhorn-manager/datastore"
+	"github.com/rancher/longhorn-manager/types"
+	apiv1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/kubernetes/pkg/controller"
+
+	longhorn "github.com/rancher/longhorn-manager/k8s/pkg/apis/longhorn/v1alpha1"
+	lhfake "github.com/rancher/longhorn-manager/k8s/pkg/client/clientset/versioned/fake"
+	lhinformerfactory "github.com/rancher/longhorn-manager/k8s/pkg/client/informers/externalversions"
+
+	. "gopkg.in/check.v1"
+)
+
+type DisasterRecoveryTestCase struct {
+	volume *longhorn.Volume
+	pv     *apiv1.PersistentVolume
+	pvc    *apiv1.PersistentVolumeClaim
+	pod    *apiv1.Pod
+	node   *apiv1.Node
+	va     *storagev1.VolumeAttachment
+}
+
+func generateDisasterRecoveryTestCase() *DisasterRecoveryTestCase {
+	volume := newVolume(TestVolumeName, 2)
+	pv := newPV()
+	pvc := newPVC()
+	pod := newPodWithPVC()
+	node := newKubernetesNode(TestNode1, apiv1.ConditionUnknown, apiv1.ConditionUnknown,
+		apiv1.ConditionUnknown, apiv1.ConditionUnknown, apiv1.ConditionUnknown,
+		apiv1.ConditionUnknown, apiv1.ConditionUnknown)
+	va := newVA(TestVAName, TestNode1, TestPVName)
+
+	node.Status.VolumesAttached = []apiv1.AttachedVolume{
+		apiv1.AttachedVolume{
+			DevicePath: TestVAName,
+			Name:       "kubernetes.io/csi/io.rancher.longhorn^" + TestPVCName,
+		},
+	}
+	pod.Status.Phase = apiv1.PodPending
+	pvc.Status.Phase = apiv1.ClaimBound
+	volume.Status.KubernetesStatus = types.KubernetesStatus{
+		PVName:       TestPVName,
+		PVStatus:     string(apiv1.VolumeBound),
+		Namespace:    TestNamespace,
+		PVCName:      TestPVCName,
+		PodName:      TestPodName,
+		PodStatus:    string(apiv1.PodRunning),
+		WorkloadName: TestWorkloadName,
+		WorkloadType: TestWorkloadKind,
+		LastPodRefAt: getTestNow(),
+	}
+
+	return &DisasterRecoveryTestCase{
```

Comment:
Add test case, e.g. don't delete VA if the node is not down, don't delete VA if two pods using the same PVC.